src/v/config/property_constraints.cc

```diff
@@ -0,0 +1,462 @@
+// Copyright 2023 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "config/property_constraints.h"
+
+#include "cluster/types.h"
+#include "config/configuration.h"
+#include "model/fundamental.h"
+
+#include <chrono>
+#include <stdexcept>
+#include <string>
+
+namespace config {
+
+std::string_view to_string_view(constraint_type type) {
+    switch (type) {
+    case constraint_type::none:
+        return "none";
+    case constraint_type::_restrict:
+        return "restrict";
+    case constraint_type::clamp:
+        return "clamp";
+    }
+
+    return "invalid";
+}
+
+template<>
+std::optional<constraint_type>
+from_string_view<constraint_type>(std::string_view sv) {
+    return string_switch<std::optional<constraint_type>>(sv)
+      .match("none", constraint_type::none)
+      .match("restrict", constraint_type::_restrict)
+      .match("clamp", constraint_type::clamp)
+      .default_match(constraint_type::none);
+}
+
+// This is where all the type conversions happen
+std::shared_ptr<constraint_validator_t> constraint_validator_map(
+  ss::sstring& name,
+  std::optional<ss::sstring> min_str_opt,
+  std::optional<ss::sstring> max_str_opt,
+  constraint_enabled_t enabled) {
+    auto& cfg = config::shard_local_cfg();
+
+    if (name == "log_cleanup_policy") {
+        return std::make_shared<
+          constraint_validator_enabled<model::cleanup_policy_bitflags>>(
+          std::move(enabled),
+          cfg.log_cleanup_policy.bind(),
+          [](const cluster::topic_configuration& topic_cfg) {
+              return topic_cfg.properties.cleanup_policy_bitflags
+                       ? *topic_cfg.properties.cleanup_policy_bitflags
+                       : model::cleanup_policy_bitflags::none;
```

Comment:
nit I think you can rewrite this as:

`topic_cfg.properties.cleanup_policy_bitflags.value_or(model::cleanup_policy_bitflags::none)`